src/merge.cpp

```diff
@@ -0,0 +1,265 @@
+    uint32_t num_threads = vm["threads"].as<uint32_t>();
+    tbb::task_scheduler_init init(num_threads);
+    MAT::Tree mat1 = MAT::load_mutation_annotated_tree(mat1_filename);
+    MAT::Tree mat2 = MAT::load_mutation_annotated_tree(mat2_filename);
+    MAT::Tree baseMat;
+    MAT::Tree otherMat;
+    concurMap::accessor ac;
+
+    if (mat1.condensed_nodes.size() > 0)
+    {
+        mat1.uncondense_leaves();
+    }
+    if (mat2.condensed_nodes.size() > 0)
+    {
+        mat2.uncondense_leaves();
+    }
+    //Assigns biggest MAT to baseMat and smaller one to otherMat
+    if (mat1.get_num_leaves() > mat2.get_num_leaves())
+    {
+        baseMat = mat1;
+        otherMat = mat2;
+    }
+    else
+    {
+        baseMat = mat2;
+        otherMat = mat1;
+    }
+    //Checks for consistency
+
+    if (consistent(baseMat, otherMat) == false)
+    {
+        return;
+    }
+    //Makes a copy of the baseMat files to add the samples to later
+    MAT::Tree finalMat = get_tree_copy(baseMat);
+    std::vector<std::string> samples;
+    auto otherLeaves = otherMat.get_leaves_ids();
+    auto baseLeaves = baseMat.get_leaves_ids();
+    sort(otherLeaves.begin(), otherLeaves.end());
+    sort(baseLeaves.begin(), baseLeaves.end());
+
+    //creates vector of new samples to be added
+    std::set_difference(otherLeaves.begin(), otherLeaves.end(), baseLeaves.begin(), baseLeaves.end(), std::back_inserter(samples));
+
+    //concurMap::accessor ac;
+    static tbb::affinity_partitioner ap;
+     std::cout<<samples.size()<<std::endl;
+
+    tbb::mutex m;
+    //parallel loop that concurrently goes through all samples
+    tbb::parallel_for(
+        tbb::blocked_range<size_t>(0, samples.size()),
+        [&](tbb::blocked_range<size_t> r)
+        {
+            for (size_t k = r.begin(); k < r.end(); k++)
+            {
+
+                std::unordered_map<int, std::string> anc;
+
+                std::string x = samples[k];
+                auto ancestors = otherMat.rsearch(x, true);
+                std::string curr = finalMat.root->identifier;
+                std::vector<int> indices;
+                for (int y = 0; y < ancestors.size(); y++)
+                {
+                    m.lock();
+                    if (consistNodes.find(ac, ancestors[y]->identifier) == true)
+                    {
+
+                        anc[y] = ac->second;
+                        indices.push_back(y);
+
+                        ac.release();
+                    }
+                    m.unlock();
+                }
+
+                int min;
+
+                if (indices.size() != 0)
+                {
+                    min = *std::min_element(indices.begin(), indices.end());
+                    curr = anc[min];
+                }
+                else
+                {
+                    min = anc.size() - 1;
+                }
+                m.lock();
+                finalMat.create_node(x, curr, -1);
```

Comment:
Check out mapper2_body (https://github.com/yatisht/usher/blob/282552a180773b62a51cd6faa70d3054b2afd706/src/usher_mapper.cpp#L167) and how it is being used by usher (https://github.com/yatisht/usher/blob/282552a180773b62a51cd6faa70d3054b2afd706/src/usher.cpp#L626). Instead of adding samples to the `curr` node, you should find the most parsimonious node descendant of `curr` (get the list of descendants through a BFS expansion rooted at `curr`). This should be quite fast, I think, if you carefully parallelize it.